atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/UpdateExecutor.java

```diff
@@ -55,8 +60,24 @@ public void update(Cell cell, long ts, byte[] oldValue, byte[] newValue) {
         int updated = ((PalantirSqlConnection) conns.get()).updateCountRowsUnregisteredQuery(sqlString,
                 args);
         if (updated == 0) {
-            // right now we don't know what's actually in the db :-(
-            throw new CheckAndSetException(cell, tableRef, oldValue, ImmutableList.of());
+            throw new CheckAndSetException(cell, tableRef, oldValue, getCurrentValues(cell, ts, prefixedTableName));
         }
     }
+
+    private List<byte[]> getCurrentValues(Cell cell, long ts, String prefixedTableName) {
+        String sqlString = "/* SELECT (" + prefixedTableName + ") */"
+                + " SELECT val FROM " + prefixedTableName
+                + " WHERE row_name = ?"
+                + " AND col_name = ?"
+                + " AND ts = ?";
+        Object[] args = new Object[] {
+                cell.getRowName(),
+                cell.getColumnName(),
+                ts
+        };
+        AgnosticLightResultSet results = conns.get().selectLightResultSetUnregisteredQuery(sqlString, args);
```

Comment:
You can't run "UPDATE, then SELECT if the UPDATE failed" in a single transaction, because when the update fails, the transaction will get rolled back. We don't want to do the SELECT first, because in most cases that would be a pointless DB round trip.